homeassistant/components/vesync/__init__.py

```diff
@@ -1 +1,116 @@
-"""The vesync component."""
+"""Etekcity VeSync integration."""
+import logging
+import voluptuous as vol
+from pyvesync import VeSync
+from homeassistant.const import CONF_USERNAME, CONF_PASSWORD
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.config_entries import SOURCE_IMPORT
+from .common import (async_process_devices, CONF_SWITCHES)
+from .config_flow import configured_instances
+from .const import (DOMAIN, VS_DISPATCHERS, VS_DISCOVERY,
+                    SERVICE_UPDATE_DEVS, CONF_MANAGER)
+
+_LOGGER = logging.getLogger(__name__)
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+async def async_setup(hass, config):
+    """Set up the VeSync component."""
+    conf = config.get(DOMAIN)
+
+    if conf is None:
+        return True
+
+    if isinstance(conf, dict) and not configured_instances(hass):
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={'source': SOURCE_IMPORT},
+                data={
+                    CONF_USERNAME: conf[CONF_USERNAME],
+                    CONF_PASSWORD: conf[CONF_PASSWORD]
+                }))
+
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set up Vesync as config entry."""
+    username = config_entry.data[CONF_USERNAME]
+    password = config_entry.data[CONF_PASSWORD]
+
+    time_zone = str(hass.config.time_zone)
+
+    manager = VeSync(username, password, time_zone)
+
+    login = await hass.async_add_executor_job(manager.login)
+
+    if not login:
+        _LOGGER.error("Unable to login to the VeSync server")
+        return False
+
+    device_dict = await async_process_devices(hass, manager)
+
+    forward_setup = hass.config_entries.async_forward_entry_setup
+
+    hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][CONF_MANAGER] = manager
+
+    switches = hass.data[DOMAIN][CONF_SWITCHES] = []
+
+    hass.data[DOMAIN][VS_DISPATCHERS] = []
+
+    if device_dict[CONF_SWITCHES]:
+        switches.extend(device_dict[CONF_SWITCHES])
+        hass.async_create_task(forward_setup(config_entry, 'switch'))
+
+    async def async_new_device_discovery(service):
+        """Discover if new devices should be added."""
+        manager = hass.data[DOMAIN][CONF_MANAGER]
+        switches = hass.data[DOMAIN][CONF_SWITCHES]
+
+        dev_dict = await async_process_devices(hass, manager)
+        switch_devs = dev_dict.get(CONF_SWITCHES, [])
+
+        if switch_devs:
+            switch_set = set(switch_devs)
+            new_switches = list(switch_set.difference(switches))
+            switch_len = len(new_switches)
+            if switch_len > 0:
+                switches.extend(new_switches)
+                if switches:
```

Comment:
Since we extend `switches` above this case will always be true if there are `new_switches` and we will never forward platform setup. This won't work.

Reply:
Thanks for the catch, I'm embarrassed :grimacing: